### ISSUE TITLE
What about a KNNQueryBuilder in Java REST Client ?

### ISSUE THREAD

**Issue by ivan4github** [alexklibisz/elastiknn#258]
ElasticKnn is a very elegant solution to search BERT Embeddings for similarities far more efficiently than with scoring only.
It is simple to install and use with a very light print on an existing cluster.
Thank you for this fantastic work.

The only thing that stops us on using it in production is that we are using Java REST Client to handle queries and retrieval.
This is where the hard part is for us.

A suggestion, ...it is only a suggestion ....

What about a **KNNQueryBuilder** in Java REST Client ?

I don't know anything about plugins and cannot tell wether it is feasible...

Thanks anyway

**Comment by alexklibisz** [comment]
Hey, thanks for the kind remarks. I will look into making a Java query builder. Hopefully elastic has provided interfaces that can be extended for this.

**Comment by alexklibisz** [comment]
@ivan4github I took a first pass at this in #260. There is a snapshot release here: https://oss.sonatype.org/content/repositories/snapshots/com/klibisz/elastiknn/client-java_2.12/. When you import this package, you'll get a class called `ElastiknnNearestNeighborsQueryBuilder`. This includes just enough functionality to execute queries through the Java Rest Client. There's a test demonstrating the functionality here: https://github.com/alexklibisz/elastiknn/pull/260/files#diff-4939334b1302b43334985c2ec6760add6bdb679d837665ab3c5e23dcd18c2e93R32-R37

**Comment by ivan4github** [comment]
Alex,

Waow, very impressive, a great functionnal response within one week end day !

The problem is that I am not lightning fast as you are....
For the future happy users in Java REST API like I am, here is the translation of your scala code in goog old java ;-)

I combined a queryString and an Angular KNN query (which looks the more appropriate to our BERT embeddings) in a single request:

`            // String query builder
            QueryStringQueryBuilder queryBuilder = QueryBuilders.queryStringQuery(query).defaultField("tout").defaultOperator(Operator.AND).quoteFieldSuffix(".exact");

            // Elasticknn Angular Query Builder
            float[] myTestList = new float[3];
            myTestList[0] = (float)0.1;
            myTestList[1] = (float)0.2;
            myTestList[2] = (float)0.3;
            Vector myTestVect = new Vector.DenseFloat(myTestList);
            ElastiknnNearestNeighborsQuery.AngularLsh eknnQuery=  new ElastiknnNearestNeighborsQuery.AngularLsh(myTestVect, 50);
            ElastiknnNearestNeighborsQueryBuilder eknnQueryBuilder = new ElastiknnNearestNeighborsQueryBuilder(eknnQuery, "Phrase_vector_knn");

            // combine both in Multi Search request (direct from Java REST Client doc)
            MultiSearchRequest multiRequest = new MultiSearchRequest();

            SearchRequest firstSearchRequest = new SearchRequest();
            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
            searchSourceBuilder.query(queryBuilder);
            firstSearchRequest.source(searchSourceBuilder);
            multiRequest.add(firstSearchRequest);

            SearchRequest secondSearchRequest = new SearchRequest();
            searchSourceBuilder = new SearchSourceBuilder();
            searchSourceBuilder.query(eknnQueryBuilder);
            secondSearchRequest.source(searchSourceBuilder);
            multiRequest.add(secondSearchRequest);`

Thank you for everything Alex.

**Comment by ivan4github** [comment]
For archives to combine the 2 query builder in one, do the following:

```
            // String query builder
            QueryStringQueryBuilder queryBuilder = QueryBuilders.queryStringQuery(query).defaultField("tout").defaultOperator(Operator.AND).quoteFieldSuffix(".exact");

            // Elasticknn Angular Query Builder
            float[] myTestList = new float[3];
            myTestList[0] = (float)0.1;
            myTestList[1] = (float)0.2;
            myTestList[2] = (float)0.3;
            Vector myTestVect = new Vector.DenseFloat(myTestList);
            ElastiknnNearestNeighborsQuery.AngularLsh eknnQuery=  new ElastiknnNearestNeighborsQuery.AngularLsh(myTestVect, 50);
            ElastiknnNearestNeighborsQueryBuilder eknnQueryBuilder = new ElastiknnNearestNeighborsQueryBuilder(eknnQuery, "Phrase_vector_knn");

                // combine 2 query builders
                QueryBuilder mqueryBuilder = QueryBuilders.boolQuery()
                    .must(queryBuilder)
                    .filter(eknnQueryBuilder);
```

**Comment by alexklibisz** [comment]
Hi @ivan4github, do you have any more input on the implementation over in #260 ?

**Comment by ivan4github** [comment]
Nothing to add, it works great.
Very easy to implement, great job !
Plase add the link to the jar for java client containing this builder for the current version.
The only hassle is the mandatory perfect match between ES version and the plugin version. A small price to pay ;-)

Please close the case.

**Comment by alexklibisz** [comment]
Great. This is released in 7.13.2.1, with some minimal docs here: https://elastiknn.com/libraries/#java-library-with-elasticsearch-query-builder-for-elastiknn-queries
